A76-ring-hash-improvements.md

```diff
@@ -0,0 +1,176 @@
+coma `,` character between each value before hashing.
+- If `request_metadata_key` is not empty, and the request has no value or an
+empty value associated with the metadata key defined, then the picker will
+generate a random hash for it. The use of a random hash key will effectively
+sends the request to a random endpoint.
+- If `request_metadata_key` is not set or empty, then the request hash key will
+be based on the xDS hash policy in RDS, which keeps the existing LB
+configuration for ring hash working as before with xDS.
+- If `request_metadata_key` is not set or empty but the xDS configuration does
+not provide the hash key, then the picker will generate a random hash for it to
+select a random endpoint, which matches the current behavior for xDS.
+
+
+### Explicitly setting the endpoint hash key
+
+The `ring_hash` policy will be changed such that the hash key used for
+determining the locations of each endpoint on the ring will be extracted from a
+pre-defined name resolver attribute called `hash_key`. If this attribute is set,
+then the endpoint is placed on the ring by hashing its value. If this attribute
+is not set or empty, then the endpoint IP address is used (current
+behavior). The location of an existing endpoint on the ring changes if its
+`hash_key` name resolver attribute changes.
+
+The xDS resolver will be changed so that when converting EDS responses to
+resolver endpoints, it will set the `hash_key` name resolver attribute to the
+value of [LbEndpoint.Metadata][LbEndpoint.Metadata] `envoy.lb` `hash_key` field,
+as described in [Envoy's documentation for the ring hash load
+balancer](envoy-ring-hash).
+
+### LB Policy Config Changes
+
+After the addition of this field, the `ring_hash` LB policy config will be:
+
+    message RingHashLoadBalancingConfig {
+      // A client-side cap will limit these values.  If either of these values
+      // are greater than the client-side cap, they will be treated as the
+      // client-side cap.  The default cap is 4096.
+      uint64 min_ring_size = 1;        // Optional, defaults to 1024.
+      uint64 max_ring_size = 2;        // Optional, defaults to 4096, max is 8M.
+
+      string request_metadata_key = 3; // Optional, defaults to the empty string.
+    }
+
+
+### Temporary environment variable protection
+
+Explicitly setting the request hash key cannot possibly lead to problem with
+existing deployment because the new behavior requires setting a load balancing
+policy configuration field that did not exist before. Therefore, it is not gated
+behind an environment variable.
+
+The second behavior change will be enabled by the
+`GRPC_EXPERIMENTAL_XDS_RING_HASH_ENDPOINT_HASH_KEY` environment variable. This
+will protect from the case where an xDS control plane is already setting the
+`LbEndpoint.Metadata` `envoy.lb` `hash_key` field, in which case deploying this
+new behavior would churn all endpoint hash keys. This environment variable will
```

Comment:
Deploying a new version of gRPC will clear all ring_hash state anyway, so I don't understand how this environment variable would do anything to prevent churn.

Reply:
Deploying with the environment variable unset will keep the current behavior of hashing the IP address instead of the new behavior of using the value from LbEndpoint.Metadata. As a result, the locations of existing endpoints on the ring will be guaranteed to be the same as before, and requests will continue being routed to the same endpoint. It will churn connections, if that's what you mean, but not endpoint locations on the ring. Suddenly changing all locations on the ring may have unintended consequences, removing the locality that ring hash may have been used for in the first place.

I'm thinking of a better phrasing, and the best I could think of was to replace "churn" with "change". Let me know if this is still unclear.